linkerd/app/core/src/transport/allow_ips.rs

```diff
@@ -0,0 +1,53 @@
+use crate::{svc, transport::OrigDstAddr};
+use std::{collections::HashSet, net::SocketAddr, sync::Arc};
+use thiserror::Error;
+
+#[derive(Clone, Debug, Default)]
+pub struct AllowIps {
+    ips: Arc<HashSet<SocketAddr>>,
+}
+
+#[derive(Clone, Debug, Error)]
+#[error("inbound connections are not allowed on this IP address ({ip})")]
+pub(crate) struct InvalidIp {
+    ip: SocketAddr,
+}
+
+impl<T> svc::stack::Predicate<T> for AllowIps
+where
+    T: svc::Param<OrigDstAddr>,
+{
+    type Request = T;
+
+    fn check(&mut self, target: T) -> Result<Self::Request, crate::Error> {
+        // Allowlist not configured.
+        if self.ips.is_empty() {
+            return Ok(target);
+        }
+
+        let OrigDstAddr(ip) = target.param();
+        if self.ips.contains(&ip) {
+            return Ok(target);
+        }
+
+        tracing::warn!(%ip, allowed = ?self.ips, "IP address not in `LINKERD2_PROXY_INBOUND_IPS`");
```

Comment:
```suggestion
        tracing::warn!(%ip, allowed = ?self.ips, "Target IP address not permitted");
```

The environment variable isn't really a user-facing concern.